src/utils/common.py

```diff
@@ -118,24 +117,89 @@ def _set_seed(seed: int):
+    criterion: torch.nn.modules.loss.CTCLoss, ocr: OCRHead, embeddings: torch.Tensor, texts: list, char2int: dict
 ):
     logits = ocr(embeddings)
     log_probs = torch.nn.functional.log_softmax(logits, dim=2)
     input_lengths = torch.LongTensor([log_probs.shape[0]] * log_probs.shape[1])
 
     chars = list("".join(np.concatenate(texts).flatten()))
-    targets = torch.LongTensor([char2int_dict.get(c, char2int_dict["UNK"]) for c in chars])
+    targets = torch.LongTensor([char2int[c] for c in chars])
 
     get_len = np.vectorize(len)
-    target_lengths = pad_sequence([torch.from_numpy(get_len(arr)) for arr in texts], batch_first=True, padding_value=0)
-
+    # target_lengths = pad_sequence([torch.from_numpy(get_len(arr)) for arr in texts], batch_first=True, padding_value=0)
+    target_lengths = []
+    for arr in texts:
+        target_lengths.append(torch.from_numpy(get_len(arr)))
+    target_lengths = torch.cat(target_lengths, dim=0)
     ctc_loss = criterion(log_probs, targets, input_lengths, target_lengths)
     ctc_loss /= len(texts)
 
     return ctc_loss
 
 
+def create_noise_mask(batch_size, seq_len, not_padded, noise_density=0.8, span_lens=6, min_unmasked=4):
+    # 1. Create random mask
+    rand_nums = torch.rand(batch_size, seq_len, device=not_padded.device)
+    mask = rand_nums.le(noise_density)
+
+    # 2. Calculate position of each mask in span
+    cum_sum = torch.cumsum(mask, dim=-1)
+    cum_sum_not_mask = torch.masked_fill(cum_sum, mask, 0)
+    mask_position = cum_sum - torch.cummax(cum_sum_not_mask, dim=-1).values
+
+    # 3. Avoid too long spans
+    min_values = torch.min(span_lens * torch.ones_like(not_padded), not_padded // 2)
+    mask = torch.where(mask_position > min_values, torch.zeros_like(mask), mask)
+
+    # 4. Ensure at least min_unmasked unmasked patches after each block
+    not_mask = ~mask
+    cum_sum_unmasked = torch.cumsum(not_mask, dim=-1)
+    cum_sum_masked = torch.masked_fill(cum_sum_unmasked, not_mask, 0)
+    unmasked_position = cum_sum_unmasked - torch.cummax(cum_sum_masked, dim=-1).values
+
+    """
+    Get indexes of the first patches of unmasked blocks. 
+    Calculate next min_unmasked-1 indexes.
+    Mark all retrieved indexes as unmasked.
+    """
+    first_unmasked_idx = torch.nonzero(unmasked_position == 1, as_tuple=True)
+    extra_idx = torch.arange(0, min_unmasked, device=unmasked_position.device).view(1, -1)
+    unmasked_spans_idx = first_unmasked_idx[1].unsqueeze(1) + extra_idx
+    unmasked_spans_idx = torch.clamp(unmasked_spans_idx, 0, unmasked_position.shape[1] - 1)
+    batch_idx = first_unmasked_idx[0].unsqueeze(1).repeat(1, min_unmasked)
+
+    unmasked_position[batch_idx.flatten(), unmasked_spans_idx.flatten()] = 1
+    mask[unmasked_position > 0] = False
+
+    return mask
+
+
+def plot_slices(
+    decoded: tuple[torch.Tensor, torch.Tensor],
+    orig: tuple[torch.Tensor, torch.Tensor],
+    iter_num: int,
+    loss: float,
+    folder_name: str = None,
+):
```

Comment:
Btw, since we are using wandb to log train polts, we can also log images
https://docs.wandb.ai/guides/track/log/media